File: Code_samples/C#/Fhi.OpenApiClient/FhiOpenApiService.cs
using System.Text;
using System.Text.Json;
using Fhi.OpenApiClient.Models;

namespace Fhi.OpenApiClient
{
    internal class FhiOpenApiService
    {
        private readonly IHttpClientFactory _httpClientFactory;
        private readonly JsonSerializerOptions _jsonSerializerOptions;
        private const string FhiOpenApiClient = "FHI-Open-API";

        public FhiOpenApiService(IHttpClientFactory httpClientFactory, JsonSerializerOptions jsonSerializerOptions)
        {
            _httpClientFactory = httpClientFactory;
            _jsonSerializerOptions = jsonSerializerOptions;
        }

        public async Task<IList<SourceOpenResponse>?> GetSourcesAsync(CancellationToken cancellationToken = default)
        {
            using var httpClient = _httpClientFactory.CreateClient(FhiOpenApiClient);
            var httpResponseMessage = await httpClient.GetAsync("/api/open/v1/Common/source", cancellationToken);

            if (!httpResponseMessage.IsSuccessStatusCode) return null;
            var responseData = await httpResponseMessage.Content.ReadAsStringAsync(cancellationToken);
            var sources = JsonSerializer.Deserialize<List<SourceOpenResponse>?>(responseData, _jsonSerializerOptions);
            return sources;
        }

        public async Task<IList<TableOpenResponse>?> GetTablesAsync(string sourceId, DateTime? modifiedAfter = null, CancellationToken cancellationToken = default)
        {
            using var httpClient = _httpClientFactory.CreateClient(FhiOpenApiClient);
            var requestUri = $"/api/open/v1/{sourceId}/table";

            if (modifiedAfter != null)
            {
                requestUri += "?modifiedAfter=" + modifiedAfter.Value.ToString("O");
            }

            var httpResponseMessage = await httpClient.GetAsync(requestUri, cancellationToken);

            if (!httpResponseMessage.IsSuccessStatusCode) return null;
            var responseData = await httpResponseMessage.Content.ReadAsStringAsync(cancellationToken);
            var tables = JsonSerializer.Deserialize<List<TableOpenResponse>>(responseData, _jsonSerializerOptions);

            return tables;
        }

        public async Task<MetadataOpenResponse?> GetMetadataAsync(string sourceId, int tableId, CancellationToken cancellationToken = default)
        {
            using var httpClient = _httpClientFactory.CreateClient(FhiOpenApiClient);
            var httpResponseMessage = await httpClient.GetAsync($"/api/open/v1/{sourceId}/table/{tableId}/metadata", cancellationToken);

            if (!httpResponseMessage.IsSuccessStatusCode) return null;

            var responseData = await httpResponseMessage.Content.ReadAsStringAsync(cancellationToken);
            var metadata = JsonSerializer.Deserialize<MetadataOpenResponse?>(responseData, _jsonSerializerOptions);

            return metadata;
        }

        public async Task<IList<FlagOpenResponse>?> GetFlagsAsync(string sourceId, int tableId, CancellationToken cancellationToken = default)
        {
            using var httpClient = _httpClientFactory.CreateClient(FhiOpenApiClient);
            var httpResponseMessage = await httpClient.GetAsync($"/api/open/v1/{sourceId}/table/{tableId}/flag", cancellationToken);

            if (!httpResponseMessage.IsSuccessStatusCode) return null;

            var responseData = await httpResponseMessage.Content.ReadAsStringAsync(cancellationToken);
            var flags = JsonSerializer.Deserialize<List<FlagOpenResponse>>(responseData, _jsonSerializerOptions);

            return flags;
        }

        public async Task<DimensionsOpenResponse?> GetDimensionsAsync(string sourceId, int tableId, CancellationToken cancellationToken = default)
        {
            using var httpClient = _httpClientFactory.CreateClient(FhiOpenApiClient);
            var httpResponseMessage = await httpClient.GetAsync($"/api/open/v1/{sourceId}/table/{tableId}/dimension", cancellationToken);

            if (!httpResponseMessage.IsSuccessStatusCode) return null;

            var responseData = await httpResponseMessage.Content.ReadAsStringAsync(cancellationToken);
            var dimensions = JsonSerializer.Deserialize<DimensionsOpenResponse>(responseData, _jsonSerializerOptions);

            return dimensions;
        }

        public async Task<DataRequest?> GetQueryAsync(string sourceId, int tableId, CancellationToken cancellationToken = default)
        {
            using var httpClient = _httpClientFactory.CreateClient(FhiOpenApiClient);
            var httpResponseMessage = await httpClient.GetAsync($"/api/open/v1/{sourceId}/table/{tableId}/query", cancellationToken);

            if (!httpResponseMessage.IsSuccessStatusCode) return null;

            var responseData = await httpResponseMessage.Content.ReadAsStringAsync(cancellationToken);
            var query = JsonSerializer.Deserialize<DataRequest>(responseData, _jsonSerializerOptions);

            return query;
        }

        public async Task<IList<string[]>?> GetDataAsync(string sourceId, int tableId, DataRequest query, CancellationToken cancellationToken = default)
        {
            if (query.Response == null || query.Dimensions == null) return null;

            query.Response.Format = "csv3";     //Supported formats are json-stat2, csv2 and csv3
            UpdateQueryToReturnAllMeasureTypesFilteredOnFirstCategory(query);

            var dataRequest = JsonSerializer.Serialize(query);
            var requestContent = new StringContent(dataRequest, Encoding.UTF8, "application/json");

            using var httpClient = _httpClientFactory.CreateClient(FhiOpenApiClient);
            var httpResponseMessage = await httpClient.PostAsync($"/api/open/v1/{sourceId}/table/{tableId}/data", requestContent, cancellationToken );

            if (!httpResponseMessage.IsSuccessStatusCode) return null;

            var csvData = new List<string[]>();
            await using var contentStream = await httpResponseMessage.Content.ReadAsStreamAsync(cancellationToken);

            using var streamReader = new StreamReader(contentStream);
            while (!streamReader.EndOfStream)
            {
                var line = await streamReader.ReadLineAsync(cancellationToken);
                if (line == null) continue;
                var row = line.Replace("\"","").Split(',');
                csvData.Add(row);
            }

            return csvData;
        }

        private void UpdateQueryToReturnAllMeasureTypesFilteredOnFirstCategory(DataRequest query)
        {
            if (query.Dimensions == null) return;
            foreach (var dimension in query.Dimensions)
            {
                if (dimension.Code == "MEASURE_TYPE")
                {
                    dimension.Filter = "all"; //Supported filters are item, all and top
                    dimension.Values = new[] {"*"};
                }
                else
                {
                    dimension.Filter = "top"; //Supported filters are item, all and top
                    dimension.Values = new[] {"1"};
                }
            }
        }
    }
}

File: Code_samples/Kotlin/FhiOpenApiClient/src/main/kotlin/FhiOpenApiService.kt

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import models.*
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse.BodyHandlers
import java.time.LocalDate

internal class FhiOpenApiService {
    private val baseURI = "https://statistikk-data.fhi.no"
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun getSourcesAsync(): List<SourceOpenResponse>? = withContext(Dispatchers.IO) {
        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseURI/api/open/v1/Common/source"))
            .GET()
            .version(HttpClient.Version.HTTP_2)
            .build()

        val response = HttpClient.newBuilder()
            .build()
            .sendAsync(httpRequest, BodyHandlers.ofString())
            .await()

        if (response.statusCode() != 200) return@withContext null

        val sources = json.decodeFromString<List<SourceOpenResponse>>(response.body())

        return@withContext sources
    }

    suspend fun getTablesAsync(sourceId: String, modifiedAfter: LocalDate? = null): List<TableOpenResponse>? = withContext(Dispatchers.IO) {
        var requestUrl = "$baseURI/api/open/v1/$sourceId/table"
        if (modifiedAfter != null)
        {
            requestUrl += "?modifiedAfter=$modifiedAfter"
        }

        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create(requestUrl))
            .GET()
            .version(HttpClient.Version.HTTP_2)
            .build()

        val response = HttpClient.newBuilder()
            .build()
            .sendAsync(httpRequest, BodyHandlers.ofString())
            .await()

        if (response.statusCode() != 200) return@withContext null

        val tables = json.decodeFromString<List<TableOpenResponse>>(response.body())

        return@withContext tables
    }

    suspend fun getMetadataAsync(sourceId: String, tableId: Int): MetadataOpenResponse? = withContext(Dispatchers.IO){
        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseURI/api/open/v1/$sourceId/table/$tableId/metadata"))
            .GET()
            .version(HttpClient.Version.HTTP_2)
            .build()

        val response = HttpClient.newBuilder()
            .build()
            .sendAsync(httpRequest, BodyHandlers.ofString())
            .await()

        if (response.statusCode() != 200) return@withContext null

        val metadata = json.decodeFromString<MetadataOpenResponse>(response.body())

        return@withContext metadata
    }

    suspend fun getFlagsAsync(sourceId: String, tableId: Int): List<FlagOpenResponse>? = withContext(Dispatchers.IO){
        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseURI/api/open/v1/$sourceId/table/$tableId/flag"))
            .GET()
            .version(HttpClient.Version.HTTP_2)
            .build()

        val response = HttpClient.newBuilder()
            .build()
            .sendAsync(httpRequest, BodyHandlers.ofString())
            .await()

        if (response.statusCode() != 200) return@withContext null

        val flags = json.decodeFromString<List<FlagOpenResponse>>(response.body())

        return@withContext flags
    }

    suspend fun getDimensionsAsync(sourceId: String, tableId: Int): DimensionsOpenResponse? = withContext(Dispatchers.IO){
        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseURI/api/open/v1/$sourceId/table/$tableId/dimension"))
            .GET()
            .version(HttpClient.Version.HTTP_2)
            .build()

        val response = HttpClient.newBuilder()
            .build()
            .sendAsync(httpRequest, BodyHandlers.ofString())
            .await()

        if (response.statusCode() != 200) return@withContext null

        val dimensions = json.decodeFromString<DimensionsOpenResponse>(response.body())

        return@withContext dimensions
    }

    suspend fun getQueryAsync(sourceId: String, tableId: Int): DataRequest? = withContext(Dispatchers.IO){
        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseURI/api/open/v1/$sourceId/table/$tableId/query"))
            .GET()
            .version(HttpClient.Version.HTTP_2)
            .build()

        val response = HttpClient.newBuilder()
            .build()
            .sendAsync(httpRequest, BodyHandlers.ofString())
            .await()

        if (response.statusCode() != 200) return@withContext null

        val query = json.decodeFromString<DataRequest>(response.body())

        return@withContext query
    }

    suspend fun getDataAsync(sourceId: String, tableId: Int, query: DataRequest): List<List<String>>? = withContext(Dispatchers.IO){
        if (query.response == null || query.dimensions == null) return@withContext null

        query.response!!.format = "csv3" //Supported formats are json-stat2, csv2 and csv3
        updateQueryToReturnAllMeasureTypesFilteredOnFirstCategory(query)

        val body = json.encodeToString(query)

        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseURI/api/open/v1/$sourceId/table/$tableId/data"))
            .headers("Content-Type", "application/json;charset=UTF-8")
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .version(HttpClient.Version.HTTP_2)
            .build()

        val response = HttpClient.newBuilder()
            .build()
            .sendAsync(httpRequest, BodyHandlers.ofString())
            .await()

        if (response.statusCode() != 200) return@withContext null

        val data = response.body()

        val csvData = mutableListOf<List<String>>()

        val inputStream = data.byteInputStream()
        val reader = BufferedReader(InputStreamReader(inputStream))

        while (true) {
            val line = reader.readLine() ?: break
            if (line.isEmpty()) continue

            val lineSplit = line.replace("\\", "").split(",")
            csvData.add(lineSplit)
        }
        reader.close()

        return@withContext csvData
    }

    private fun updateQueryToReturnAllMeasureTypesFilteredOnFirstCategory(query: DataRequest) {
        if (query.dimensions == null) return
        for (dimension in query.dimensions!!) {
            if (dimension.code == "MEASURE_TYPE") {
                dimension.filter = "all" // Supported filters are item, all and top
                dimension.values = arrayOf("*")
            } else {
                dimension.filter = "top" // Supported filters are item, all and top
                dimension.values = arrayOf("1")
            }
        }
    }
}